### ISSUE TITLE
New check for self/cls assignments doesn't catch all possible cases

### ISSUE THREAD

**Issue by skirpichev** [PyCQA/pylint#2100]
### Steps to reproduce
1.
```
ylint) sk@note:~ $ cat b.py 
class B:
    def foo(self, bar):
        def spam(x):
            nonlocal self
            self = x
        spam(bar)
        print(self)

b = B()
b.foo(42)
```
2. 
```
(pylint) sk@note:~ $ pylint -d all -e W0642 b.py 
No config file found, using default configuration

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```
3. 
```
(pylint) sk@note:~ $ python3 b.py 
42
```
### Current behavior

No warning on step 2.  But clearly, self was assigned.

### Expected behavior

W0642 warning
